src/notebooks/controllers/preferredKernelConnectionService.ts

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { CancellationToken, NotebookDocument } from 'vscode';
+import { ContributedKernelFinderKind, IContributedKernelFinder } from '../../kernels/internalTypes';
+import { PreferredRemoteKernelIdProvider } from '../../kernels/jupyter/preferredRemoteKernelIdProvider';
+import {
+    IKernelFinder,
+    KernelConnectionMetadata,
+    LiveRemoteKernelConnectionMetadata,
+    LocalKernelSpecConnectionMetadata,
+    PythonKernelConnectionMetadata,
+    RemoteKernelConnectionMetadata,
+    RemoteKernelSpecConnectionMetadata
+} from '../../kernels/types';
+import { disposeAllDisposables } from '../../platform/common/helpers';
+import { IDisposable } from '../../platform/common/types';
+import { getNotebookMetadata, translateKernelLanguageToMonaco } from '../../platform/common/utils';
+import { IInterpreterService } from '../../platform/interpreter/contracts';
+import { ServiceContainer } from '../../platform/ioc/container';
+import { getLanguageOfNotebookDocument } from '../languages/helpers';
+
+/**
+ * Attempt to clean up https://github.com/microsoft/vscode-jupyter/issues/11914
+ * Provides the ability to look for exact or preferred kernel connections.
+ * Note: This class has zero side effects unlike `ControllerPreferredService`, and is meant to be a replacement for it
+ */
+export class PreferredKernelConnectionService {
+    private readonly disposables: IDisposable[] = [];
+    public dispose() {
+        disposeAllDisposables(this.disposables);
+    }
+    public async findExactRemoteKernelConnection(
```

Comment:
@rebornix @Tyriar 
this is a very simple approach to not injecting everything
i personally think making everything as injectable and injecting all classes instead if instantiating then when we need them is not absolutely necessary, but that depends on whether one believes everything must be injected or not. i don't think we'd need to

Reply:
Do you mean fetching the `kernelFinder`? Removing injection would end up making code less readable wouldn't it, reducing code quality in the end?

Reply:
> Removing injection would end up making code less readable wouldn't it, reducing code quality in the end?

Not sure about readability, but we're still using a DI container, instead of injection we're using a service container to pull in the necessary dependencies as and when we need them.
Happy to change this back to an injectable.

Reply:
I.e. we still use the DI container, but don't inject, instead use the service container to create/get what we need.

Reply:
This can help us with the startup since inversify doesn't need to resolve the dependency graph for it. Though as @Tyriar mentioned it can potentially make code less readable. For `IKernelFinder` using the DI to resolve the service other than injectable seems identical, for others I think we still need to go through them and see if they should just be a local instance other than a global service.

Reply:
I'm not following, are you suggesting I change this?
If so, what's the difference between this approach and the use of `IInstantiationService.createInstance` in VS Code?
Just trying to understand the use case for the latter in VS Code, so I can follow the same pattern in this repo (without having to worry about any issues with maintainability).

Reply:
@DonJayamanne sorry for the confusion. There are two scenarios:

In this case, `KernelFinder` might be used in mulitple places, so it makes good sense to register it to `ServiceContainer` but not `injectable`. And when we want to use it, we do `ServiceContainer.get`.

The other one is: Service A depends on service B, service B is only used in Service A and it's not required in `ctor`. Service B has its own service dependencies. When we need to use an instance of service B, then we would use `IInstantiationService.createInstance` to create an instance of service B with the right services dependencies.

With this approach, Service B is never registered on ServiceContainer, it won't be available to anyone else.